cyclic_boosting/generic_loss.py

```diff
@@ -0,0 +1,678 @@
+
+    Returns
+    -------
+    float
+        calcualted quantile costs
+    """
+    if not len(y) > 0:
+        raise ValueError("Loss cannot be computed on empty data")
+    else:
+        sum_weighted_error = np.nansum(
+            ((y < prediction) * (1 - quantile) * (prediction - y) + (y >= prediction) * quantile * (y - prediction))
+            * weights
+        )
+        return sum_weighted_error / np.nansum(weights)
+
+
+def quantile_global_scale(X, y, quantile, weights, prior_prediction_column, link_func):
+    """
+    Calculation of the global scale for quantile regression, corresponding
+    to the (continuous approximation of the) respective quantile of the
+    target values used in the training.
+
+    The exact value of the global scale is not critical for the model
+    accuracy (as the model has enough parameters to compensate), but a
+    value not representating a good overall average leads to factors with
+    averages unequal to 1 for each feature (making interpretation more
+    difficult).
+    """
+    if weights is None:
+        raise RuntimeError("The weights have to be initialized.")
+
+    global_scale_link_ = link_func(continuous_quantile_from_discrete(y, quantile))
+
+    prior_pred_link_offset_ = None
+    if prior_prediction_column is not None:
+        prior_pred = get_X_column(X, prior_prediction_column)
+        finite = np.isfinite(prior_pred)
+        if not np.all(finite):
+            _logger.warning(
+                "Found a total number of {} non-finite values in the prior prediction column".format(np.sum(~finite))
+            )
+
+        prior_pred_mean = np.sum(prior_pred[finite] * weights[finite]) / np.sum(weights[finite])
+
+        prior_pred_link_mean = link_func(prior_pred_mean)
+
+        if np.isfinite(prior_pred_link_mean):
+            prior_pred_link_offset_ = global_scale_link_ - prior_pred_link_mean
+        else:
+            warnings.warn(
+                "The mean prior prediction in link-space is not finite. "
+                "Therefore no indiviualization is done "
+                "and no prior mean substraction is necessary."
+            )
+            prior_pred_link_offset_ = float(global_scale_link_)
+
+    return global_scale_link_, prior_pred_link_offset_
+
+
+def model_multiplicative(param, yhat_others):
+    return param * yhat_others
+
+
+def model_additive(param, yhat_others):
+    return param + yhat_others
+
+
+def uncertainty_gamma(y, weights):
+    # use moment-matching of a Gamma posterior with a log-normal
+    # distribution as approximation
+    alpha_prior = 2
+    alpha_posterior = np.sum(y) + alpha_prior
+    sigma = np.sqrt(np.log(1 + alpha_posterior) - np.log(alpha_posterior))
+    return sigma
+
+
+def uncertainty_gaussian(y, weights):
+    return np.sqrt(np.mean(y) / len(y))
+
+
+def uncertainty_beta(y, weights, link_func):
+    # use moment-matching of a Beta posterior with a log-normal
+    # distribution as approximation
+    alpha_prior = 1.001
```

Comment:
To avoid having multiple identical constant values related to the beta distribution, we could put `get_beta_factors` (see [link](https://github.com/Blue-Yonder-OSS/cyclic-boosting/blob/main/cyclic_boosting/classification.py#L23)) in `utils.py` and use this here, unless we later want to change these parameters in one place but not the other.

Reply:
I keep it in classification.py and import that here.